chainer_chemistry/models/relgcn.py

```diff
@@ -0,0 +1,186 @@
+    """RelGCN submodule for readout part
+
+    Args:
+        in_channels (int): dimension of feature vector associated to
+            each atom (node)
+        out_channels (int): output dimension of feature vector
+            associated to each molecule (graph)
+        nobias (bool): If ``True``, then this function does not use
+            the bias.
+    """
+
+    def __init__(self, in_channels, out_channels, nobias=True):
+        super(RelGCNReadout, self).__init__()
+        with self.init_scope():
+            self.sig_linear = chainer_chemistry.links.GraphLinear(
+                in_channels, out_channels, nobias=nobias)
+            self.tanh_linear = chainer_chemistry.links.GraphLinear(
+                in_channels, out_channels, nobias=nobias)
+
+    def __call__(self, h, x=None):
+        """
+        (implicit:
+            N is number of edges, R is number of types of relations)
+        Args:
+            h: (batchsize, num_nodes, ch)
+                N x F : Matrix of edges, each row is a molecule and
+                each column is a feature.
+                F_l is the number of features at layer l
+                F_0, the input layer, feature is type of molecule.
+                Softmaxed
+
+            x: (batchsize, num_nodes, ch)
+
+        Returns:
+            (batchsize, ch)
+            F_n : Graph level representation
+
+        Notes:
+            I think they just incorporate "no edge" as one of the
+            categories of relations, i've made it a separate tensor
+            just to simplify some implementation, might change later
+        """
+        if x is None:
+            in_feat = h
+        else:
+            in_feat = functions.concat([h, x], axis=2)
+        sig_feat = functions.sigmoid(self.sig_linear(in_feat))
+        tanh_feat = functions.tanh(self.tanh_linear(in_feat))
+
+        return functions.tanh(functions.sum(sig_feat * tanh_feat, axis=1))
+
+
+class RelGCN(chainer.Chain):
+
+    """Relational GCN (RelGCN)
+
+    See: Michael Schlichtkrull+, \
+        Modeling Relational Data with Graph Convolutional Networks. \
+        March 2017. \
+        `arXiv:1703.06103 <https://arxiv.org/abs/1703.06103>`
+
+    Args:
+        out_channels (int): dimension of output feature vector
+        num_edge_type (int): number of types of edge
+        ch_list (list): channels of each update layer
+        n_atom_types (int): number of types of atoms
+        input_type (str): type of input vector
+        scale_adj (bool): If ``True``, then this network normalizes
+            adjacency matrix
+    """
+
+    def __init__(self, out_channels=64, num_edge_type=4, ch_list=None,
+                 n_atom_types=MAX_ATOMIC_NUM, input_type='int',
+                 scale_adj=False):
+
+        super(RelGCN, self).__init__()
+        ch_list = ch_list or [16, 128, 64]
+        # ch_list = [in_channels] + ch_list
+        with self.init_scope():
+            if input_type == 'int':
```

Comment:
This is a question. Do we need this argument? In GGNN, it was inferred by input dtype. https://github.com/pfnet-research/chainer-chemistry/blob/master/chainer_chemistry/models/ggnn.py#L124
Is it different from GGNN situation?
How do you think? @corochann